src/combinatorics/set_family/compression/uv.lean

```diff
@@ -0,0 +1,240 @@
+/-
+Copyright (c) 2021 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import data.finset.basic
+
+/-!
+# UV-compressions
+
+This file defines UV-compression. It is an operation on a set family that reduces its shadow.
+
+UV-compressing `a : α` along `u v : α` means replacing `a` by `(a ⊔ u) \ v` if `a` and `u` are
+disjoint and `v ≤ a`. In some sense, it's moving `a` from `v` to `u`.
+
+UV-compressions are immensely useful to prove the Kruskal-Katona theorem. The idea is that
+compressing a set family might decrease the size of its shadow, so iterated compressions hopefully
+minimise the shadow.
+
+## Main declarations
+
+* `uv.compress`: `compress u v a` is `a` compressed along `u` and `v`.
+* `uv.compression`: `compression u v s` is the compression of the set family `s` along `u` and `v`.
+  It is the compressions of the elements of `s` whose compression is not already in `s` along with
+  the element whose compression is already in `s`. This way of splitting into what moves and what
+  does not ensures the compression doesn't squash the set family, which is proved by
+  `uv.card_compress`.
+
+## Notation
+
+`𝓒` (typed with `\MCC`) is notation for `uv.compression` in locale `finset_family`.
+
+## Notes
+
+Even though our emphasis is on `finset α`, we define UV-compressions more generally in a generalized
+boolean algebra, so that one can use it for `set α`.
+
+## TODO
+
+Prove that compressing reduces the size of shadow. This result and some more already exist on the
+branch `combinatorics`.
+
+## References
+
+* https://github.com/b-mehta/maths-notes/blob/master/iii/mich/combinatorics.pdf
+
+## Tags
+
+compression, UV-compression, shadow
+-/
+
+open finset
+
+variable {α : Type*}
+
+/-- UV-compression is injective on the elements it moves. See `uv.compress`. -/
+lemma sup_sdiff_inj_on [generalized_boolean_algebra α] (u v : α) :
```

Comment:
Compression is not in general injective, right? We could have the two `inj_on`s for both sides of the if.

Reply:
No, compression isn't injective. What do you mean by "both sides of the if"? `{x | ¬ (disjoint u x ∧ v ≤ x)}.inj_on (λ x, (x ⊔ u) \ v)`? This is pretty useless to us as `(x ⊔ u) \ v` won't be the compression of `x` if `¬ (disjoint u x ∧ v ≤ x)`. Instead, its compression will be itself.
Actually, I don't even think it's true.

Reply:
I was thinking of two lemmas: [the element it moves].`inj_on uv.compress` (which is what you just proved here, but stated in a different way). [the elements it doesn't move],`inj_on uv.compress` (as it's the identity).

Reply:
That still sounds a bit dull to me 😅

Reply:
yeah but it's basically there ;b

Reply:
i was somewhat confused why this lemma was written about the "set version" instead of `compress`, because even if you want it more generally you need to import this file

Reply:
I already wasn't in favor of adding that `sup_sdiff_inj_on` lemma, so I'll skip my turn here.

Reply:
that's fair